### PR TITLE
add center loss

### PR DESCRIPTION
**Pull request by ShownX**
Implement Center Loss in Gluon, tested individually.

### CODE REVIEW

File: python/mxnet/gluon/loss.py [hunk 1]
```diff
@@ -696,3 +696,35 @@ def hybrid_forward(self, F, pred, positive, negative):
...
+    def hybrid_forward(self, F, output, label, sample_weight=None):
+        selected_centers = F.take(self._centers, label)
+        diff = output - selected_centers
+        loss = F.sum(F.square(diff), axis=1)
+
+        loss = _apply_weighting(F, loss, self._weight, sample_weight)
+        # Update the centers
+        counts = nd.zeros((self._num_classes, ), ctx=self._ctx)
+        for lb in label:
+            lb = int(lb.asscalar())
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: These are bad practice. We should add a bincount operator to do this

File: python/mxnet/gluon/loss.py [hunk 2]
```diff
@@ -696,3 +696,35 @@ def hybrid_forward(self, F, pred, positive, negative):
...
+        self._centers = nd.zeros((num_classes, feature_size), ctx=ctx)
+
+    def hybrid_forward(self, F, output, label, sample_weight=None):
+        selected_centers = F.take(self._centers, label)
+        diff = output - selected_centers
+        loss = F.sum(F.square(diff), axis=1)
+
+        loss = _apply_weighting(F, loss, self._weight, sample_weight)
+        # Update the centers
+        counts = nd.zeros((self._num_classes, ), ctx=self._ctx)
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: using nd means this is not a hybridblock anymore

File: python/mxnet/gluon/loss.py [hunk 3]
```diff
@@ -381,3 +381,35 @@ def hybrid_forward(self, F, data, label,
                                  data_lengths=data_lengths, label_lengths=label_lengths,
                                  blank_label='last')
         return _apply_weighting(F, loss, self._weight, sample_weight)
+
+
+class CenterLoss(Loss):
+    r""" Center Loss
```
[the remark this answers is not in the document]
Comment (ShownX) [reply python/mxnet/gluon/loss.py]: Will do

File: python/mxnet/gluon/loss.py (hunk 1 above)
Comment (ShownX) [reply python/mxnet/gluon/loss.py]: Will change.

File: python/mxnet/gluon/loss.py (hunk 2 above)
Comment (ShownX) [reply python/mxnet/gluon/loss.py]: Sorry, I did not get it. @szha suggested updating the centers manually. What should I use in this case?

Comment (szha) [reply python/mxnet/gluon/loss.py]: What are you referring to? Here if you explicitly use ndarray, the hybridized version or the symbolic version of training won't work properly, hence it's discouraged to use ndarray-specific functionality in HybridBlocks. Loss class is a child class of HybridBlock.

Comment (ShownX) [reply python/mxnet/gluon/loss.py]: So how can update the centers for each class? using the ```params```?

Comment (szha) [reply python/mxnet/gluon/loss.py]: Declaring a parameter is the usual way of keeping states in operators. I'm not sure about what way would be appropriate for the update though.

**Comment by ShownX** [comment]
changes made here: https://github.com/ShownX/mxnet-center-loss. Closed.
